### ISSUE TITLE
Mega.nz Session Cache

### ISSUE THREAD

**Issue by metabox-cloud** [rclone/rclone#4695]
#### What is your current rclone version (output from `rclone version`)?
rclone v1.53.1

#### What problem are you are trying to solve?

` 2020/10/19 02:35:08 ERROR : *FILENAME*.rar: Failed to copy: upload file failed to create session: Object (typically, node or user) not found`

#### How do you think rclone should be changed to solve that?
I have contacted Mega.nz support, which is trying to get me to go to megacmd, although all my stuff is rclone based.. so don't really want to, and they have informed me that rclone doesn't cache the session on mega. which causes multiple logins (one for every file) and in part, gets limited in the end.. especially if running multiple threads over multiple servers..

![ZTYX5FUhLL](https://user-images.githubusercontent.com/72478288/96390345-ba00a500-11fb-11eb-90a3-9f5e093ca53b.png)

**Comment by ImportTaste** [comment]
Fuck, no wonder. So this is what's been going wrong.

**Comment by ncw** [comment]
Rclone logs in each time it is run, so if you running rclone lots then you'll get lots of logins.

However rclone does work with the webdav server in megacmd quite well. It isn't quite as fast as using rclone directly but it works pretty well.

**Comment by ImportTaste** [comment]
My current workaround, on Windows, is mounting my MEGA with a batch script...

```batch
start "rcd" cmd /c rclone rcd --rc-no-auth
rclone rc mount/mount fs=MEGA: mountPoint=C:/mount/MEGA "vfsOpt={\"CacheMode\": 2, \"CaseInsensitive\": false}"
```

And then performing operations on the mount instead, using `rclone rc vfs/forget` to refresh the mounted file system cache as needed.

**Comment by ncw** [comment]
Yes, running a mount (or `rclone serve webdav` etc) will also make rclone persistent.

**Comment by ImportTaste** [comment]
> Rclone logs in each time it is run, so if you running rclone lots then you'll get lots of logins.
> 
> However rclone does work with the webdav server in megacmd quite well. It isn't quite as fast as using rclone directly but it works pretty well.

I tried doing what you suggested here, using megacmd's webdav server, but apparently this is prone to IO errors like these:
```
2020/11/16 12:48:44 ERROR : IO error: couldn't list files: XML syntax error on line 42: invalid character entity &percnt;
2020/11/16 12:50:00 ERROR : IO error: couldn't list files: XML syntax error on line 253: invalid character entity &add;
2020/11/16 12:50:00 ERROR : IO error: couldn't list files: XML syntax error on line 253: invalid character entity &add;
2020/11/16 12:50:00 ERROR : IO error: couldn't list files: XML syntax error on line 253: invalid character entity &add;
```

**Comment by toddfries** [comment]
instead of mounting webdav as a fs, I simply am transferring data to the webdav using rclone from other cloud providers.

what is discussed above doesn't quite give the full picture however.

MEGA session is a full copy of all files in the account sent to the login .. and it is intended to be persistent.  The mega command itself caches this metadata locally, and so listings and such are very fast. If desired, one can sign out of the session with the megacmd, which exits the background daemon it spawns to maintain the session.  It persists across reboots, and stays signed in until you sign out, like a never expiring cookie in your web browser, for example.

On the other hand, when rclone signs in, it creates a new session for each connection, and never signs out.

So I was having issues with mega not letting me login because I had too many sessions.

There is a place in their website you can go to remove sessions, as well as via the megacmd.

Bottom line for me is, if I use rclone to sign in / out (aka overnight sync some files up to mega) it is something I'd have to automate cleaning up (perhaps megacmd can let me script clearing the sessions..).

If rclone were to do one thing to help all mega users, it would be to figure out how to sign out of the sessions when it exits.

Perhaps a session could be a sign in process then stored in the rclone config similar to how tokens are handled, so at least it would re-use sessions vs creating new ones each time it starts.

Thanks!

**Comment by CodingKoopa** [comment]
> I tried doing what you suggested here, mounting megacmd's webdav server, but apparently this is prone to IO errors like these:
> 
> ```
> 2020/11/16 12:48:44 ERROR : IO error: couldn't list files: XML syntax error on line 42: invalid character entity &percnt;
> 2020/11/16 12:50:00 ERROR : IO error: couldn't list files: XML syntax error on line 253: invalid character entity &add;
> 2020/11/16 12:50:00 ERROR : IO error: couldn't list files: XML syntax error on line 253: invalid character entity &add;
> 2020/11/16 12:50:00 ERROR : IO error: couldn't list files: XML syntax error on line 253: invalid character entity &add;
> ```

I have gotten to the bottom of this as much as I can, and documented my findings in meganz/sdk#2599. The gist is, MegaCMD is escaping all characters with HTML4 entities, but the XML parser is not recognizing them. One solution is to remove all of these HTML4 mappings from `sdk/src/utils.cpp`, except for the [5 entities that are built into XML](https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Predefined_entities_in_XML). This could also potentially be rectified on rclone's end, though, if its XML parser recognized these entities. I can see that Go's [`encoding/xml`](https://pkg.go.dev/encoding/xml) package is used around the codebase, but I didn't investigate further than that.

**Comment by CodingKoopa** [comment]
I have went through the different methods of accessing MEGA within rclone and documented the issues and successes I had. What follows immediately is a lengthy description of my process, and then a summary of my findings. I deliberated for some time over the best way to organize this; unfortunately, I am not sure how to make this easier to parse without losing information. 

<details>

## Setup

### Software

#### rclone
My rclone version:
```
rclone v1.58.0-DEV
- os/version: arch rolling (64 bit)
- os/kernel: 5.15.12-arch1-1 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17.5
- go/linking: dynamic
- go/tags: none
```
I built my binary from commit b55575e6225f7b297ea7636272705855aae5486f. Using the latest development build was necessary to integrate #5848.

In rclone, I have the relevant remotes setup as follows:
- My `local` remote is setup to abide by Windows character limitations (`encoding = Slash,LtGt,DoubleQuote,Colon,Question,Asterisk,Pipe,BackSlash,Ctl,RightSpace,RightPeriod,InvalidUtf8,Dot`), as some of the directories are shared with a Windows system.
- My `mega` remote is an instance of the `mega` backend, connected to my MEGA account.
- My `mega-webdav` remote is an instance of the `webdav` backend, connected to the locally hosted WebDAV share.

#### MEGAcmd
My MEGAcmd version:
```
MEGAcmd version: 1.4.0.0: code 1040000
```
To automatically have MEGAcmd serve my MEGA files over WebDAV, I created and enabled a [user systemd service](https://gitlab.com/CodingKoopa/comet-observatory/-/blob/7beeef119a47d6fd56208cfef47f0a5283cf9294/config/systemd-user-units/mega-cmd-server.service).

It's worth noting that MEGAcmd 1.4.0 is over a year old. Since then, a [1.4.1 tag for Windows](https://github.com/meganz/MEGAcmd/releases/tag/1.4.1_Win) has been created. In further experimentation, using the latest development build may be worth trying.

## Process

### Syncing and deduping with `mega`
To reset the state as much as possible, I:
- Synchronize `local` with `mega`.
  ```sh
  rclone sync /home/koopa/Documents mega:Documents --verbose --progress --links --delete-excluded --filter-from=/home/koopa/Documents/Projects/Bash/comet-observatory/config/rclone-filter-list.ini
  ```
  The [filter list](https://gitlab.com/CodingKoopa/comet-observatory/-/blob/fcdef050b705ae3391486ff0e7d9f7245cfa3906/config/rclone-filter-list.ini) shouldn't be anything too interesting.
- Dedupe `mega`.
  ```sh
  rclone dedupe --dedupe-mode newest mega: --verbose --progress
  ```

In doing so, however, I did notice that `rclone dedupe` removed a necessary file which ended up having to be reuploaded:
```
$ rclone dedupe --dedupe-mode newest mega: --verbose --progress
2022-01-04 19:54:40 INFO  : mega root '': Looking for duplicate names using newest mode.
...
2022-01-04 19:54:46 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects: Merging contents of duplicate directories
2022-01-04 19:54:46 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects: removing empty directory
2022-01-04 19:54:47 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects/pack: Merging contents of duplicate directories
2022-01-04 19:54:47 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects/pack: removing empty directory
2022-01-04 19:54:48 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects/pack: merging "pack-b2caaee813576d0740cc11ff1af249f8fe6e2def.idx"
2022-01-04 19:54:49 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects/pack: merging "pack-b2caaee813576d0740cc11ff1af249f8fe6e2def.pack"
2022-01-04 19:54:49 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects/pack: removing empty directory
2022-01-04 19:54:50 INFO  : Documents/Projects/Python/BDSP-Randomizers/.git/objects/pack: removing empty directory
Transferred:              0 B / 0 B, -, 0 B/s, ETA -
Elapsed time:        59.8s
$ rclone sync /home/koopa/Documents mega:Documents --verbose --progress --links --delete-excluded --filter-from=/home/koopa/Documents/Projects/Bash/comet-observatory/config/rclone-filter-list.ini
2022-01-04 19:55:46 INFO  : Projects/Python/BDSP-Randomizers/.git/objects/pack/pack-b2caaee813576d0740cc11ff1af249f8fe6e2def.idx: Copied (new)
2022-01-04 20:33:31 INFO  : Projects/Python/BDSP-Randomizers/.git/objects/pack/pack-b2caaee813576d0740cc11ff1af249f8fe6e2def.pack: Copied (new)
```
I didn't touch the directory in question, and I don't think I changed anything in MEGA, so I'm not sure why this happened.

Starting the MEGAcmd WebDAV share, it still sees a duplicate directory. On the left is mega.nz, and on the right is the WebDAV share viewed in Firefox, seemingly out of sync, despite refreshing the page.
![Screenshot_20220104_204607](https://user-images.githubusercontent.com/13039555/148324663-f5413ecb-b308-47d2-a779-ab9f62d4dc01.png)

Restarting the MEGAcmd server fixes this, which is a little weird because I started the server only after the rclone operations finished.

### Syncing with `mega-webdav`
With the WebDAV share still running, I began testing it with rclone, with the same command but with the new remote:
```
rclone sync /home/koopa/Documents mega-webdav:Documents --verbose --progress --links --delete-excluded --filter-from=/home/koopa/Documents/Projects/Bash/comet-observatory/config/rclone-filter-list.ini
```

In commands, assume the working directory is `/home/koopa/Documents` on `local`.

Initializing:
- Ran a sync. All is well (`There was nothing to transfer`).

Adding an empty new file:
- Created the new file:
  ```sh
  touch new_file
  ```
- Ran another sync. The file was successfully copied (`Copied (new)`). The file is indeed shown on the WebDAV share viewed in Firefox.

Removing a file:
- Removed the new file:
  ```sh
  rm new_file
  ```
- Ran another sync. The file was successfully deleted (`Deleted`).

Modifying a file:
- Redid *Adding an empty new file*.
- Modified the file:
  ```sh
  echo "Hello world!" > new_file
  ```
- Ran another sync. The file was said to be replaced (`Copied (replaced existing)`)
- In the WebDAV Share, there are now two of `new_file`.
  - This was identified using both the share accessed from Firefox, and using `rclone lsf`.
  - In the actual MEGA account, there indeed is a duplicate.
- Removed the file:
  ```sh
  rm new_file
  ```
- Ran another sync, file was deleted (`Deleted`) with the original remaining:
  ```
  2022-01-05 01:37:42 NOTICE: new_file: Duplicate object found in destination - ignoring
  2022-01-05 01:39:17 INFO  : new_file: Deleted
  ```
  A failure has occurred somewhere along the way.

Adding a large new file:
- Created the new file. This file is a larger one (162MiB), and one which happened to cause an issue earlier.
  ```sh
  cp Projects/Python/BDSP-Randomizers/.git/objects/pack/pack-b2caaee813576d0740cc11ff1af249f8fe6e2def.pack .
  ```
- Ran another sync. The transfer was seemingly completed in a couple of minutes, but wouldn't be committed:
  ```
  Transferred:      323.842 MiB / 323.842 MiB, 100%, 6.183 MiB/s, ETA 0s
  Checks:             53727 / 53727, 100%
  Transferred:            0 / 1, 0%
  Elapsed time:      5m10.6s
  Transferring:
   * pack-b2caaee813576d074…f1af249f8fe6e2def.pack:100% /161.921Mi, 20.238Mi/s, 0s
  ```
  I interrupted it via Ctrl+C. The file is not present in the WebDAV share.
- Ran another sync. It hangs in the same way, completing the same amount of checks.
- Arbitrarily made the file smaller (162MiB -> 50 MiB):
  ```
  dd if=pack-b2caaee813576d0740cc11ff1af249f8fe6e2def.pack of=pack-b2caaee813576d0740cc11ff1af249f8fe6e2def.pack bs=50MiB count=1
  ```
- Ran another sync. The file is successfully copied right away, and the transfer is successful. A failure occurred in transferring the larger file, it seems.

Retrieving a file:
- Copied a file from `mega-webdav`:
  ```
  rclone copy mega-webdav:/Documents/file .
  ```
  This works without issues.

In addition to these usecases, there have been less reproducible oddities I've seen:
- Issues with encoding:
  ```
  2021/12/06 16:39:17 ERROR : ProgramConfigurations/Dolphin/Profiles/GCPad/Keyboard - Mario Kart%3a Double Dash!!.ini: Failed to copy: object not found
  2021/12/06 16:39:17 INFO  : ProgramConfigurations/Dolphin/Profiles/GCPad/Keyboard - Mario Kart. Double Dash!!.ini: Copied (new)
  ```
  This file started out with a `:` in the name, and eventually in MEGA it became the case that there existed both that file and a duplicate with the `%3a` entity. Locally, I renamed it to use a `.` insteaf of `:`, and, as demonstrated here, it was still a mess. I ultimately deleted the cursed file, and don't intend on following up on this one.
- Issues with `rclone dedupe` not finding nodes on `mega`:
  ```
  2021/12/26 19:50:20 ERROR : merge duplicate dirs: MergeDirs failed to find node for: Documents/Projects/C#/PKHeX/PKHeX.Core/Legality/Areas
  ```
  I do have captured a full rclone log, but I have had issues reproducing this, and I think it might have arisen from a great number of duplicates existing.

### Persisting rclone by mounting
Mounting `mega`:
- Mounted the remote (where `./mega` is a local directory):
  ```
  rclone mount mega: ./mega
  ```
  I was able to browse the directory in Dolphin without issues, though the following warning was issued (trimmed at `<SNIP>` over potential security concerns).
  ```
  2022/01/05 21:40:10 Unsolicited response received on idle HTTP channel starting with "\xb8<SNIP>"; err=<nil>
  ```

Mounting `mega-webdav`:
- Mounted the remote:
  ```
  rclone mount mega-webdav: ./mega-webdav
  ```
  I was able to browse the mountpoint without issues.

Though syncing with the "persistent" `mega` mountpoint theoretically seems like another method for syncing with, syncing with a mountpoint for a cloud services as if it's a local directory doesn't seem like a great idea.

### Persisting rclone and controlling remotely
For this section, I started a remote control instance of rclone:
```
rclone rcd --rc-no-auth
```

Accessing `mega`:
- Executed a directory listing:
  ```
  rclone rc core/command command=lsd -a mega:
  ```
- Repeated the same command. It appears to take the same amount of delay, with each invokation taking around 15s (~20s in worser cases). Even after invoking the command 20+ times, I was not rate limited by MEGA, confirming that staying logged in is the key.

Accessing `mega-webdav`:
- Executed a directory listing:
  ```
  rclone rc core/command command=lsd -a mega-webdav:
  ```
  This works, and is an order of magnitude faster than `mega`.

Syncing using `mega`:
- Ran the equivalent of the sync command from earlier:
  ```
  rclone rc core/command command=sync -a "/home/koopa/Documents" -a "mega:Documents" -o verbose -o progress -o links -o delete-excluded -o filter-from="/home/koopa/Documents/Projects/Bash/comet-observatory/config/rclone-filter-list.ini"
  ```
  This wasn't accepted as correct syntax for `rclone sync`, and I couldn't figure out how to make it work.
</details>

#### rclone Concerns
- `mega` backend:
  - This backend appears to be rather slow. `lsd` on the root directory takes no less than 13s, whereas with the `webdav` backend via MEGAcmd it takes less than 1s.
  - `rclone dedupe` functionality doesn't appear to be perfect, with it appearing to delete a file without leaving any non-duplicates. Doesn't seem to occur in most circumstances.
  - When there are duplicates, `MergeDirs failed to find node` can occur. Seems to occur when there are great amounts of duplicates.
- `webdav` with MEGAcmd:
  - **When `rclone sync` intends to update an existing file, a new file is created of the same name.** Duplicate files will proliferate over time.
  - Large (>=100MiB maybe?) file transactions appear to hang after 100%.

I don't expect that these are all rclone issues; there are surely problems elsewhere in the stack that are causing or exacerbating these symptoms.

#### MEGAcmd Concerns
- It appears to be possible for the WebDAV listing to become desynchronized with the MEGA account. Doesn't seem to occur in most circumstances.
  - rclone users should probably be advised against mixing these methods, in any case.
- In Firefox, navigating to a `C#` folder of the share brings me to the `C` folder that happens to be in the same directory. The URL is C#, but MEGAcmd doesn't seem to understand this.
  - Fortunately, rclone has no issues with navigating the `C#` directory.

#### Takeaway
- `mega` is relatively slow, and has rate limiting issues severe enough to discourage users from using it as-is.
- `webdav` with MEGAcmd is fast, but has too many issues currently. Specifically, this shouldn't be recommended to users without addresing the bolded issue above.
- `mega`, remotely controlled, still feels slow, but at least is stable and shouldn't have issues.

Though I couldn't figure out the syntax for my `sync` command, I think that the `mega` backend, remotely controlled, is the best working option as-is.

**Comment by ImportTaste** [comment]
@CodingKoopa The bolded issue refers to only syncing TO the WebDAV rather than syncing FROM the WebDAV, yes? I'm mainly concerned with syncing from it, so I don't think it would affect me, but I just want to be sure I understand correctly.

**Comment by CodingKoopa** [comment]
I think that's correct. The symptom is that, when rclone goes to replace an existing file on the WebDAV server, an additional file is created. If you were going in the other direction, and rclone found itself in the situation of having to update an existing local file, I think it would do just fine.

(in other words, if you are syncing *to* local (or elsewhere) *from* WebDAV, that's a different scenario than what I tested yeah)

**Comment by ImportTaste** [comment]
> I think that's correct. The symptom is that, when rclone goes to replace an existing file on the WebDAV server, an additional file is created. If you were going in the other direction, and rclone found itself in the situation of having to update an existing local file, I think it would do just fine.
> 
> (in other words, if you are syncing _to_ local (or elsewhere) _from_ WebDAV, that's a different scenario than what I tested yeah)

Have you tried using the `--delete-before` flag? That, ideally, should make it so that the file is deleted on MEGA so that the duplicate does not occur.

**Comment by CodingKoopa** [comment]
I'm not sure if there's even a notion of a file deletion in this transaction. As far as I understand, the `--delete-(before,during,after)` option allows you to control *when* the deletion occurs. Here it seems like rclone isn't on the same page with how the server handles it writing to an existing filename.

**Comment by ImportTaste** [comment]
> I'm not sure if there's even a notion of a file deletion in this transaction. As far as I understand, the `--delete-(before,during,after)` option allows you to control _when_ the deletion occurs. Here it seems like rclone isn't on the same page with how the server handles it writing to an existing filename.
> 
> (this is consistent with your edit)

Yeah, there'd have to be a new sync option to delete a mismatch before uploading a replacement.
